### DISCUSSION TITLE
Conditional mapping of nested objects

### DISCUSSION THREAD

**Discussion by jovan996** [mapstruct/mapstruct#3273] [Q&A]
Hello everyone,

I need small help. I have the following case:
I need to map properties from UpdateUser model to User model. UserUpdate model has roleId property and User has nested Role object which has id property. Something like this...

```java
@Getter
@SuperBuilder
public class User {

  protected UUID id;
  protected String email;
  protected UserStatus status;
  protected Role role;
  protected String firstName;
  protected String lastName;
  protected LocalDate dateOfBirth;
}
```

```java
@Getter
@Builder
public class Role {

  private Short id;
  private String name;
  private Boolean deletable;
java
```

And my mapper method:
```java
  @Mapping(target = "email", ignore = true)
  @Mapping(target = "status", ignore = true)
  @Mapping(target = "id", source = "userId")
  @Mapping(target = "role.id", source = "model.roleId")
  User fromUpdateUser(UUID userId, UpdateUser model);
```

I have a problem because I need to map model.roleId property only if it is roleId !=null and if roleId is null I need to set whole Role object as null(not just the id). The biggest problem is to set Role object to null if model.roleId is null. Does anyone know how to do this conditional mapping?

**Comment by dkellenb** [comment]
Hi @jovan996 . What is the type of the `UpdateUser.roleId`. Have you tried to create an additional mapper like:

```java
@Mapping(target = "id", source = ".")
Role fromRoleId(Short id);
```

**Reply by jovan996** [reply]
Hi @dkellenb. Thank you so much for your answer. The type for UpdateUser.roleId is also Short, so your proposal works very well. I solved this with default method which is exactly the same as generated implementation of your mapping method, but i didn't know how to write this like you in one line. Thank you once again.

**Comment by ZsergUA** [comment]
I have similar problem, but my nested property is a collection. So additional abstract mapping method will not help. Any suggestions how it can be resolved without explicit coding?

**Reply by dkellenb** [reply]
Hi @ZsergUA . Please add an example what you'd like to achieve.

**Reply by ZsergUA** [reply]
This will describe the problem. I have a lot of POJOs like A1 and A2.

```java
@Mapper
public interface MapA1toA2 {

    @Getter @Setter
    class A1 {
        private B1List blist;
        private C1List clist;
    }
    @Getter
    class B1List {
        private final List<B1> b = new ArrayList<>();
    }
    @Getter @Setter
    class B1 {
        private String id;
    }
    @Getter
    class C1List {
        private final List<C1> c = new ArrayList<>();
    }
    @Getter @Setter
    class C1 {
        private String id;
    }

    @Getter @Setter
    class A2 {
        private List<B2> b;
        private List<C2> c;

    }
    @Getter @Setter
    class B2 {
        private String id;
    }
    @Getter @Setter
    class C2 {
        private String id;
    }

    @Mapping(target = "blist.b", source = "b")
    @Mapping(target = "clist.c", source = "c")
    A1 map(A2 source);
}
```
And here is expected test to pass:
```java
class MapA1toA2ImplTest {

    private final MapA1toA2 mapper = Mappers.getMapper(MapA1toA2.class);

    @Test
    void mapEmpty() {
        A1 a = mapper.map(new A2());
        Assertions.assertNull(a.getBlist());
        Assertions.assertNull(a.getClist());
    }
}
```

**Reply by dkellenb** [reply]
Just create a mapper with following definition:

```java
@Mapper
public interface MapA1toA2 {

    A1 map(A2 source);

    @Mapping(target = "b", source = ".")
    B1List map(List<B2> source);

    @Mapping(target = "c", source = ".")
    C1List map(List<C2> source);

    B1 map(B2 source);

    C1 map(C2 source);

}
```

**Reply by ZsergUA** [reply]
And as I mentioned earlier this does not work:
`java: Can't generate mapping method from iterable type from java stdlib to non-iterable type.`
